cockatrice/src/game/player/player.cpp

```diff
@@ -1330,16 +1331,19 @@ void Player::actMoveTopCardsToExile()
 
 void Player::actMoveTopCardsUntil()
 {
-    moveTopCardTimer->stop();
-    movingCardsUntil = false;
+    stopMoveTopCardsUntil();
+
     QString expr = previousMovingCardsUntilExpr;
+    int numberOfHits = previousMovingCardsUntilNumberOfHits;
     for (;;) {
-        bool ok;
-        expr = QInputDialog::getText(game, "Put top cards on stack until", "Card name (or search expressions)", {},
-                                     expr, &ok);
-        if (!ok) {
+        DlgMoveTopCardsUntil dlg(game, expr, numberOfHits);
+        if (!dlg.exec()) {
             return;
         }
+
+        expr = dlg.getExpr();
+        numberOfHits = dlg.getNumberOfHits();
+
         movingCardsUntilFilter = FilterString(expr);
         if (movingCardsUntilFilter.valid()) {
```

Comment:
because we now have a custom dialog, the validation can now be moved into the dialog itself, that way we don't have to reopen the dialog when the user enters an invalid filter and we can show the error in the dialog

Reply:
as an aside, is it possible for the validation to ever fail? Currently (in both release and master branch cockatrice), if I try to put in an invalid search expression, it'll happily accept that (and then proceed to flip the entire deck onto the stack)

Reply:
@RickyRister Might want to add your own "fail" case where you pre-search the deck, and if you find 0 hits mark it as failed? Not sure on this one

Reply:
Wouldn't that give the player more information than they're legally entitled to, in cases where the contents of the deck aren't known (for example, if cards were exiled face-down)?

Reply:
Sure, but if they enter an invalid filter and flip the entire deck... it's kind of the same issue wouldn't you say?

Reply:
Here is an example: 
- you have a single card in your starting deck that matches the expression
-  some number of cards are exiled face down from your deck
- You enter the search expression
- The client does a pre-search of the deck
- The client now knows the full contents of the deck
- The client sees that there’s a matching card and starts flipping

The client now knows the full contents of the deck, even though legally it should only know the contents of the deck down to the matching card. This is relevant due to the face down exiled cards, as that means you can now discern the identity of those cards with 100% certainty.

Reply:
Anyways, seeing as the expression validation for this action doesn’t work on release cockatrice either, we should probably open a separate issue for fixing that.

Reply:
> @RickyRister Might want to add your own "fail" case where you pre-search the deck, and if you find 0 hits mark it as failed? Not sure on this one

this really is not possible in the current way this works, this function is very simple, it does not use any new protocol, it simply moves a card to the stack, waits for the next card to be moved to the stack, does a comparison with it, then repeats if necessary